feat: actions

antsgar opened the pull request:
- Added react-native-bottom-sheet dependency
- Created new BottomSheet component
- Created NoteBottomSheet component for sheet that opens when long pressing on a note on the list (same component should be reusable for bottom sheet on note view)
- Added react-native-svg and react-native-svg-transformer dependencies to display SVG icons
- Added new v4 SVG icons
- Added Listed actions and action handling to note bottom sheet

mobitar commented:
(Will defer to @baptiste-grob)

baptiste-grob commented on src/screens/Notes/NoteCell.tsx:
```diff
@@ -191,61 +87,73 @@ export const NoteCell = ({
+        onLongPress={onLongPress}
+      >
+        <Container
+          ref={elementRef as any}
```
Is there no way to make a ref with the correct type signature here?

antsgar replied:
Ah I think autofixing with eslint added this for me, will fix it. Thanks! 😄

antsgar replied:
Ah actually this was old code, it just looks new because of the new indentation, but will fix it now!

baptiste-grob commented on src/screens/Notes/NoteCell.tsx:
```diff
@@ -191,61 +87,73 @@ export const NoteCell = ({
+      <NoteBottomSheet
+        note={note}
+        bottomSheetRef={bottomSheetRef}
+        listedExtensions={listedExtensions}
+        loadListedExtension={loadListedExtension}
```
I don't see `listedExtensions` or `loadListedExtension` being used in NoteCell - perhaps they could be 'private' members of `NoteBottomSheet` instead, to push some complexity further down the component tree?

antsgar replied:
Sorry, could you explain this one a bit more? You're right the only thing NoteCell is doing with them is passing them down to its child NoteBottomSheet but not sure how we could change that

antsgar replied:
`listedExtensions` is part of NoteCell's state because it needs to be set when long pressing the note. `loadListedExtension` on the other hand can definitely be moved to NoteBottomSheet

baptiste-grob reviewed (commented):
Very cool!

As a general question, you seem to be using `useMemo` and `useCallback` a ton in this PR, has that yielded better performance or is it more of a systematic thing? I'm slightly concerned about readability and the cost of using these optimizations being greater then not using them.

mobitar commented on src/components/BottomSheet.tsx:
```diff
@@ -0,0 +1,408 @@
+      onDismiss={onDismiss}
+    >
+      <>
+        {title ? (
+          <TitleContainer onLayout={onTitleLayout}>
+            <Title>{title}</Title>
+          </TitleContainer>
+        ) : null}
```
Maybe `title?.length > 0`?

baptiste-grob replied:
Good point!

antsgar replied:
An empty string is falsy though so we shouldn't need to check for length? But can definitely add it if it's for better readability.

baptiste-grob replied:
`'' && 'thing'` will return `''`, which would make RN crash saying a string needs to be wrapped in a `<Text>` component

antsgar replied:
Yeah but I'm not using && here, I'm using ? which would render null in the '' case

antsgar commented:
> Very cool!
> 
> As a general question, you seem to be using `useMemo` and `useCallback` a ton in this PR, has that yielded better performance or is it more of a systematic thing? I'm slightly concerned about readability and the cost of using these optimizations being greater then not using them.

That's a very good point, I'll take a closer look!

antsgar commented on src/components/Icon.tsx:
```diff
@@ -0,0 +1,43 @@
+import IconTrash from '@Style/Icons/ic-trash.svg';
+import React from 'react';
+
+// eslint-disable-next-line no-shadow
+export enum IconType {
```
All enums trigger this eslint warning and have this eslint disable added on mobile, I haven't figured out why

antsgar replied:
Figured it out, we need to use @typescript-eslint/no-shadow rule instead of no-shadow

antsgar commented:
@baptiste-grob @mobitar thanks for your feedback! I've set this PR as draft since I still have to do some more work on it, will update soon taking your comments into account 🙂

antsgar commented on src/components/Icon.tsx:
```diff
@@ -0,0 +1,43 @@
+type Props = {
+  type: IconType;
+  size: number;
+  color: string;
```
I can't think of any easy way to do this, but might be missing something 🤔

antsgar commented on src/components/BottomSheet.tsx:
```diff
@@ -0,0 +1,408 @@
+      return [1];
+    }
+    return contentHeight < maxLimit ? [contentHeight] : [maxLimit];
+  }, [contentHeight]);
```
After testing, seems we can't drop the useMemo here since we need to ensure that the snapPoints are recalculated after the list height changes, dropping it will cause the bottom sheet to sometimes not show up after dismissing it. Also the library docs show use of useMemo for the snap points, so added it back.

antsgar commented:
@baptiste-grob removed some of the useCallback and useMemo hooks for better readability, applied changes from your comments, fixed some bugs and moved the bottom sheet up to NoteList instead of having one bottom sheet per each note cell. This means we will only have one bottom sheet regardless of the quantity of notes and should be much more performant. Feel free to review again whenever you have some time 🙂 although it might need some more readjustments as we keep testing

antsgar commented:
> @baptiste-grob removed some of the useCallback and useMemo hooks for better readability, applied changes from your comments, fixed some bugs and moved the bottom sheet up to NoteList instead of having one bottom sheet per each note cell. This means we will only have one bottom sheet regardless of the quantity of notes and should be much more performant. Feel free to review again whenever you have some time 🙂 although it might need some more readjustments as we keep testing

@baptiste-grob scratch this, there's still one issue to resolve so moved this one back to draft
